pkg/reconciler/pipelinerun/pipelinerun_test.go

```diff
@@ -1426,6 +1490,496 @@ func TestReconcileOnCancelledPipelineRun(t *testing.T) {
+		tb.PipelineRunSpec("test-pipeline", tb.PipelineRunServiceAccountName("test-sa"),
+			tb.PipelineRunCancelledRunFinally,
+		),
+		tb.PipelineRunStatus(tb.PipelineRunStartTime(time.Now()),
+			tb.PipelineRunTaskRunsStatus("test-pipeline-run-cancelled-run-finally-hello-world", &v1beta1.PipelineRunTaskRunStatus{
+				PipelineTaskName: "hello-world-1",
+				Status:           &v1beta1.TaskRunStatus{},
+			}),
+			tb.PipelineRunTaskRunsStatus("test-pipeline-run-cancelled-run-finally-final-task", &v1beta1.PipelineRunTaskRunStatus{
+				PipelineTaskName: "final-task-1",
+				Status:           &v1beta1.TaskRunStatus{},
+			}),
+		),
+	)}
+	ps := []*v1beta1.Pipeline{tb.Pipeline("test-pipeline", tb.PipelineNamespace("foo"), tb.PipelineSpec(
+		tb.PipelineTask("hello-world-1", "hello-world"),
+		tb.FinalPipelineTask("final-task-1", "some-task"),
+	))}
+	ts := []*v1beta1.Task{
+		tb.Task("hello-world", tb.TaskNamespace("foo")),
+		tb.Task("some-task", tb.TaskNamespace("foo")),
+	}
+	trs := []*v1beta1.TaskRun{
+		tb.TaskRun("test-pipeline-run-cancelled-run-finally-hello-world",
+			tb.TaskRunNamespace("foo"),
+			tb.TaskRunOwnerReference("kind", "name"),
+			tb.TaskRunLabel(pipeline.GroupName+pipeline.PipelineLabelKey, "test-pipeline-run-cancelled-run-finally"),
+			tb.TaskRunLabel(pipeline.GroupName+pipeline.PipelineRunLabelKey, "test-pipeline"),
+			tb.TaskRunSpec(tb.TaskRunTaskRef("hello-world"),
+				tb.TaskRunServiceAccountName("test-sa"),
+			),
+		),
+		tb.TaskRun("test-pipeline-run-cancelled-run-finally-final-task",
+			tb.TaskRunNamespace("foo"),
+			tb.TaskRunOwnerReference("kind", "name"),
+			tb.TaskRunLabel(pipeline.GroupName+pipeline.PipelineLabelKey, "test-pipeline-run-cancelled-run-finally"),
+			tb.TaskRunLabel(pipeline.GroupName+pipeline.PipelineRunLabelKey, "test-pipeline"),
+			tb.TaskRunSpec(tb.TaskRunTaskRef("some-task"),
+				tb.TaskRunServiceAccountName("test-sa"),
+			),
+		),
+	}
+	cms := getConfigMapsWithEnabledAlphaAPIFields()
+
+	d := test.Data{
+		PipelineRuns: prs,
+		Pipelines:    ps,
+		Tasks:        ts,
+		TaskRuns:     trs,
+		ConfigMaps:   cms,
+	}
+	prt := newPipelineRunTest(d, t)
+	defer prt.Cancel()
+
+	wantEvents := []string{
+		"Normal Started",
+	}
+	reconciledRun, clients := prt.reconcileRun("foo", "test-pipeline-run-cancelled-run-finally", wantEvents, false)
+
+	if reconciledRun.Status.CompletionTime != nil {
+		t.Errorf("Expected a CompletionTime to be nil on incomplete PipelineRun but was %v", reconciledRun.Status.CompletionTime)
+	}
+
+	// This PipelineRun should still be complete and unknown, and the status should reflect that
+	if !reconciledRun.Status.GetCondition(apis.ConditionSucceeded).IsUnknown() {
+		t.Errorf("Expected PipelineRun status to be complete and unknown, but was %v", reconciledRun.Status.GetCondition(apis.ConditionSucceeded))
+	}
+
+	// There should be exactly one task run triggered for the "final-task-1" final task
+	if len(reconciledRun.Status.TaskRuns) != 2 {
```

Comment:
The comment and the check does not match.

Also, wre we expecting two `taskRuns` here? The pipeline has one final task which has associated taskRun already running. And I think we are testing the final `taskRun` is not impacted and continues running, right?. There can not be one more taskRun. The `dag` can not be running after finally starts executing.

Reply:
Thanks for pointing that. The comment now explains that one task run represents already completes task and the second one the running finally task.